File: safesock/cpp/examples/basic_usage.cpp
// =============================================================================
// ESSENTIAL PROCESS:
// Demonstration client showcasing C++ SafeSocket connection and messaging.
//
// DATA FLOW:
// 1. Input: Target server address and message payload.
// 2. Logic: Establishes SafeSocket connection and writes framed data.
// 3. Output: Console output confirming message transmission.
//
// KEY PARAMETERS:
// - None
// =============================================================================

#include "../SafeSocket.hpp"
#include <iostream>
#include <vector>
#include <string>

int main() {
    try {
        // Attempt to create a socket. 
        // We use auto_connect=false so it doesn't try to connect immediately 
        // and fail if no server is running.
        auto sock = safesock::create("demo", "localhost:8080", "", "client", false);
        std::cout << "Socket created successfully" << std::endl;

        // We can't really test much more without a running server,
        // but this confirms the library is loaded and basic calls work.
        
        sock->close();
        std::cout << "Socket closed" << std::endl;
    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << std::endl;
        return 1;
    }
    return 0;
}

File: safesock/rust/examples/basic_usage.rs
// =============================================================================
// ESSENTIAL PROCESS:
// Basic demonstration example showcasing Rust SafeSocket connection usage.
//
// DATA FLOW:
// 1. Input: Target server address string and message bytes.
// 2. Logic: Connects via libsafesocket and sends framed payloads.
// 3. Output: Console status of socket transmission.
//
// KEY PARAMETERS:
// - None
// =============================================================================

use safesocket::SafeSocket;

fn main() {
    let lib_name = if cfg!(target_os = "macos") {
        "libsafesocket.dylib"
    } else if cfg!(target_os = "windows") {
        "libsafesocket.dll"
    } else {
        "libsafesocket.so"
    };

    // Assuming we run from safesock/rust
    let lib_path = format!("../libsafesocket/{}", lib_name);
    
    match SafeSocket::new("tcp-client", "localhost:8080", None, "client", false, &lib_path) {
        Ok(_sock) => {
            println!("SafeSocket created successfully");
        }
        Err(e) => {
            // It might fail with "connection refused" or similar if no server,
            // but "unknown profile" would also prove it works.
            eprintln!("Result: {}", e);
        }
    }
}
